broker.go

```diff
@@ -324,11 +326,30 @@ func (b *Broker) cacheMetadata(resp *proto.MetadataResp) {
 			dest := topicPartition{topic.Name, part.ID}
 			b.metadata.endpoints[dest] = part.Leader
 			debugmsg = append(debugmsg, dest, part.Leader)
+
+			count, ok := b.metadata.partitions[topic.Name]
+			if !ok || part.ID >= count {
+				b.metadata.partitions[topic.Name] = part.ID + 1
+			}
```

Comment:
This should be done outside of this loop. Is checking current value necessary? Can the number of partitions returned be lower than cached value?

```go
b.metadata.partitions[topic.Name] = len(topic.Partitions)
```